zellij-server/src/screen.rs

```diff
@@ -1537,6 +1542,97 @@ impl Screen {
         }
     }
 
+    pub fn move_active_tab_to_left(&mut self, client_id: ClientId) -> Result<()> {
+        if self.tabs.len() < 2 {
+            debug!("cannot move tab to left: only one tab exists");
+            return Ok(());
+        }
+        let Some(client_id) = self.client_id(client_id) else {
+            return Ok(());
+        };
+        let Some(&active_tab_idx) = self.active_tab_indices.get(&client_id) else {
+            return Ok(());
+        };
+        let left_tab_idx = self.index_on_left_of(active_tab_idx);
+        self.switch_tabs(active_tab_idx, left_tab_idx, client_id);
+        self.log_and_report_session_state()
+            .context("failed to move tab to left")?;
+        Ok(())
```

Comment:
Not sure if I get what you mean. But if I got it right, then I can't do that this way because:

- to calculate target index in step 1. I need `active_tab_idx`, so I'm checking it before
- to get `active_tab_idx` in the first place, via `self_active_tab_indices.get(&client_id)`, I need `client_id`, so I'm checking it before

To make the checks of `client_id` and `active_tab_index` DRY I can make `move_active_tab(client_id: ClientID, direction: Direction)` , get rid of `move_active_tab_right` and `move_active_tab_left` and then based on the `direction` passed in, calculate the target tab index. Something like this:

```
pub fn move_active_tab(&mut self, client_id: ClientId, dir: Direction) -> Result<()> {
        if dir.is_vertical() {
            panic!("cannot move tab vertically");
        }
        if self.tabs.len() < 2 {
            debug!("cannot move tab: only one tab exists");
            return Ok(());
        }
        let Some(client_id) = self.client_id(client_id) else {
            return Ok(());
        };
        let Some(&active_tab_idx) = self.active_tab_indices.get(&client_id) else {
            return Ok(());
        };

        let target_idx = if dir == Direction::Left {
            // wraps around: [tab1, tab2, tab3] => [tab1, tab2, tab3]
            //                 ^                                 ^
            //          active_tab_idx                     target_idx
            (active_tab_idx + self.tabs.len() - 1) % self.tabs.len()
        } else {
            // wraps around: [tab1, tab2, tab3] => [tab1, tab2, tab3]
            //                             ^          ^
            //                     active_tab_idx   target_idx
            (active_tab_idx + 1) % self.tabs.len()
        };

        self.switch_tabs(active_tab_idx, target_idx, client_id);
        self.log_and_report_session_state()
            .context("failed to move active tab")?;
        Ok(())
    }
```

Reply:
Actually, Yes I was thinking exactly this one. However, I think it would be better to postpone this refactoring until later.
Btw, Thank you for sharing your idea!